cvat/apps/engine/tests/utils.py

```diff
@@ -533,3 +597,244 @@ def filter_dict(
+
+
+def compare_objects(
+    self: TestCase,
+    obj1: Any,
+    obj2: Any,
+    ignore_keys: Collection[str],
+    *,
+    fp_tolerance: float = 0.001,
+    current_key: list[str] | str | None = None,
+    check_order: bool | OrderStrategy = True,
+) -> bool | NoReturn:
+    if isinstance(current_key, str):
+        current_key = [current_key]
+    elif not current_key:
+        current_key = []
+
+    key_info = f"{_format_key(current_key)}: "
+    error_msg = "{}{} != {}"
+
+    if isinstance(obj1, dict):
+        self.assertTrue(isinstance(obj2, dict), error_msg.format(key_info, obj1, obj2))
+        for k in (obj1.keys() | obj2.keys()) - set(ignore_keys):
+            compare_objects(
+                self,
+                obj1[k],
+                obj2[k],
+                ignore_keys,
+                current_key=current_key + [k],
+                fp_tolerance=fp_tolerance,
+                check_order=check_order,
+            )
+    elif isinstance(obj1, list):
+        self.assertTrue(isinstance(obj2, list), error_msg.format(key_info, obj1, obj2))
+        self.assertEqual(
+            len(obj1),
+            len(obj2),
+            error_msg.format(key_info, pformat(obj1, compact=True), pformat(obj2, compact=True)),
+        )
+
+        if check_order == True or check_order(current_key):
+            for v1, v2 in zip(obj1, obj2):
+                compare_objects(
+                    self,
+                    v1,
+                    v2,
+                    ignore_keys,
+                    current_key=current_key,
+                    fp_tolerance=fp_tolerance,
+                    check_order=check_order,
+                )
+        else:
+
+            def _compare(a, b) -> bool:
+                try:
+                    compare_objects(
+                        self,
+                        a,
+                        b,
+                        ignore_keys,
+                        current_key=current_key,
+                        fp_tolerance=fp_tolerance,
+                        check_order=check_order,
+                    )
+                    return True
+                except AssertionError:
+                    return False
+
+            _, a_unmatched, b_unmatched = _match_lists(obj1, obj2, distance=_compare)
+
+            if a_unmatched or b_unmatched:
+                self.fail(
+                    "Failed to match lists. "
+                    + error_msg.format(
+                        key_info, pformat(obj1, compact=True), pformat(obj2, compact=True)
+                    )
+                )
+
+    elif isinstance(obj1, float) or isinstance(obj2, float):
+        self.assertAlmostEqual(obj1, obj2, delta=fp_tolerance, msg=current_key)
+    else:
+        self.assertEqual(obj1, obj2, msg=current_key)
+
+
+def check_annotation_response(
+    self: TestCase,
+    response: dict,
+    data: dict,
+    *,
+    expected_source: str | None = None,
+    ignore_keys: list[str] = ["id", "version"],
```

Comment:
```suggestion
    ignore_keys: list[str] = ("id", "version"),
```
Using mutable objects as function defaults is unsafe